feat(managesieve): add XOAUTH2 authentication mechanism

felixauringer opened the pull request:
As the managesieve server only supports plain authentication, here is a first implementation of XOAUTH2 as an additional authentication mechanism for managesieve.

I would be happy about feedback :)

felixauringer committed:
feat(managesieve): add XOAUTH2 authentication mechanism

chibenwa reviewed (commented):
This looks great.

Can we try to write a test for OIDC support in managedSieve like this is done for IMAP and SMTP?

Also, out of curiosity do we have a managedSive compatible client>

We are missing:
 - ManageSiee exemple in https://github.com/apache/james-project/tree/master/examples/oidc
 - Documentation on how to configure this feature

quantranhong1999 reviewed (commented):
Read it.

Thank you for your contribution.

This lacks some tests indeed.

felixauringer committed:
fix(oidc): correct differentiation between XOAUTH2 and OAUTHBEARER

XOAUTH2 is described here: https://developers.google.com/workspace/gmail/imap/xoauth2-protocol
OAUTHBERER is described here: https://datatracker.ietf.org/doc/html/rfc5801#section-4

There is a small difference in how the user argument is called and
OAUTHBEARER also includes a GS2 header.

felixauringer committed:
feat(managesieve): add OAUTHBEARER authentication mechanism

felixauringer committed:
fix(managesieve): small fixes in authentication logic

- Authentication is now aborted when session is already authenticated
- Authentication is now aborted when client sends "*"
- Arguments by the client are now consistently unquoted
- If unquotation fails, authentication is aborted to prevent later NullPointerExceptions
- Fixed NullPointerException on AuthenticationException
- Server now sends correctly formatted response (including code) when waiting for
  additional authentication arguments
- When client sends authentication credentials in second message, the first argument
  (command) is now interpreted as argument
- When choosing mechanism failed, a proper error is returned to the client
- When argument is not quoted, authentication is aborted

I think with these changes, the server now behaves more like described in the RFC:
https://datatracker.ietf.org/doc/html/rfc5804#section-4

felixauringer committed:
test(managesieve): test authentication of managesieve server

felixauringer committed:
docs(managesieve): document usage of oidc authentication

felixauringer committed:
fix(examples): fix introspection mismatching issuer error

felixauringer committed:
feature(examples): configure managesieve and test oidc authentication in oidc example

felixauringer committed:
refactor(examples): small improvements for oidc example

- More modern syntax in compose file.
- Remove non-working links from readme.
- Use consistent container names (always ending in .example.com).
- Reduce output of test script.

felixauringer commented:
I have updated documentation and the oidc example. I added some CLI commands to test oidc authentication and made `test.sh` succeed. I did not test with thunderbird but my changes should not break the existing documentation on that.

felixauringer committed:
refactor: add more errors and reset authentication state on authentication error

While trying to get the MPT tests for managesieve running, I realized that a client
is trapped in the state "AuthenticationInProgress" when using continuation. This
made longer test scripts impossible.
To consistently reset the session, I introduced more errors that are handled in one
place instead of returning a "NO" answer directly during authentication.

Those changes also required some small changes to the MPT and the normal tests.

Additionally, there were a parsing bug when sending space-separated username and
password in the authentication continuation which is fixed now.

chibenwa commented on mpt/impl/managesieve/core/src/main/resources/org/apache/james/managesieve/scripts/capability.test:
```diff
@@ -38,8 +38,9 @@ S: "VERSION" "1.0"
-C:  user password
+S: ""
+S: OK
+C: "user password"
```
I am not fully sure that this code is compliant with eg Thunderbird extension

From my memories of SASL we either have

```
C: AUTHENTICATE "PLAIN" "AGFsaWNlAHNlY3JldA=="\r\n
S: OK "Logged in"
```

Or

```
C: AUTHENTICATE "PLAIN"\r\n
S: +\r\n
C: AGFsaWNlAHNlY3JldA==\r\n
S: OK "Logged in"\r\n
```

(with base64 being `\0LOGIN\0password`)

So seing not base64 encoded, qutoed password seems indeed weird to me...

From the above exemple the initial server response is `+ \r\n`

chibenwa commented on protocols/managesieve/src/main/java/org/apache/james/managesieve/core/PlainAuthenticationProcessor.java:
```diff
@@ -50,7 +50,7 @@ public PlainAuthenticationProcessor(UsersRepository usersRepository) {
     @Override
     public String initialServerResponse(Session session) {
-        return "+ \"\"";
+        return "\"\"\r\nOK";
```
Why do we need changing the initial server response here?

felixauringer replied:
I do not understand where this `+ ""` comes from but I didn't find it in any of the RFCs relating to ManageSieve SASL authentication.  
There is one mention in the [SASL RFC](https://www.rfc-editor.org/rfc/rfc4422#appendix-A) but only for the `EXTERNAL` mechanism which we do not use here.  
The in my opinion relevant RFC here is [RFC 5804: A Protocol for Remotely Managing Sieve Scripts](https://datatracker.ietf.org/doc/html/rfc5804#section-4). The formal syntax there defines the response to an authenticate command as follows:

```
response-authenticate = *(string CRLF)
                            ((response-ok [response-capability]) /
                             response-nobye)
                            ;; <response-capability> is REQUIRED if a
                            ;; SASL security layer was negotiated and
                            ;; MUST be omitted otherwise.
```

If we want continuation, we want to have a `response-ok` and because we do not negotiate a SASL security layer, we must omit `response-capability`.  
It then simplifies to `response-authenticate = *(string CRLF) response-ok`. Assuming we do not want to have a response code or description, it further simplifies to `response-authenticate = *(string CRLF) "OK" CRLF`.  
The [description](https://datatracker.ietf.org/doc/html/rfc5804#section-2.1) says:

> Note that an empty challenge/response is sent as an empty string.

With all our supported authentication mechanisms, the challenge sent by the server is empty, so I would argue that the server should send the empty string. Sadly, the examples in the RFC do not show this case. Using the formal syntax from above, this leads to `response-authenticate = DQUOTE DQUOTE CRLF "OK" CRLF`. (One could probably also use the literal string variant without quotes but by supplying the exact string length. However, I would argue that it does not make much sense with an empty string.)

That's the reason why I changed it. In the RFC, I do not see any possibility how an (unquoted!) `+` sign could be the response to an authenticate command. I also didn't find why this server response shouldn't end with an `OK`, but I'm less sure about that (the examples seem to omit it in the challenge-response exchanges but I don't think the formal syntax allows it). What's the reason why you use the `+` here? Is there another RFC that I do not know?

chibenwa replied:
Hello @felixauringer 

Thanks for having a look

I think as an implementer I just sticked to RFC-4616 exemples, yes (but that's10 years back.... I'm a bit rusty on the topic)

Note that the `""` is likely none compulsory and not returned eg by the IMAP layer which also implement RFC-4616...

While what comes after is subject to discussion the fact of doing a continuation is mandatory.

chibenwa commented on server/protocols/protocols-managesieve/src/test/java/org/apache/james/managesieveserver/AuthenticateTest.java:
```diff
@@ -0,0 +1,220 @@
+    // message = [authzid] UTF8NUL authcid UTF8NUL passwd
+    // The current code is more lenient and accepts the message without the first null byte.
+    @Disabled
+    @Test
```
I we think it is a good thing to be more relaxed than the RFC, I see no issue:

 - Keeping the valuable comment
 - Testing the existing behaviour with a passing test.

felixauringer committed:
fix(managesieve): Implement continuation as described in RFC-4616

felixauringer committed:
fix(oidc): wrong format of gs2-header

felixauringer committed:
refactor(managesieve): accept plain authentication without leading null byte

felixauringer committed:
feat(managesieve): add test for old SASL PLAIN parsing behavior

felixauringer commented:
> Otherall agree with proposed changes but I would need clarifications on `AUTHENTICATE "PLAIN"` handling.

Now, the behavior is the same except that the code is more strict when it comes to quoting. SASL mechanisms and the values sent by the client must be quoted according to the RFC. (In the RFC, there is the possibility of sending unquoted data which must be preceded by the length in the form `{20+}` but that was not supported before and is not supported now.)

chibenwa commented on mpt/impl/managesieve/core/src/main/resources/org/apache/james/managesieve/scripts/capability.test:
```diff
@@ -39,7 +39,7 @@ S: OK
 C: AUTHENTICATE "PLAIN"
 S: \+ ""
-C:  user password
+C: "user password"
```
When authentication happens as part of continuation I am rather sure that the crediential must nt be quoted


```suggestion
C: user password
```

CF https://datatracker.ietf.org/doc/html/rfc4616#section-4

chibenwa reviewed (commented):
Thanks for the extensive test suite it's pleasan to read.

felixauringer committed:
refactor(managesieve): only accept unquoted initial client response during continuation

chibenwa reviewed (commented):
This work looks nice to me.
I can give a hand with the picky test suite if of any use. 
Just ask.
Best regards,
Benoit

chibenwa commented:
> I can give a hand with the picky test suite if of any use.

Good job it's :green_apple: :-)

felixauringer committed:
feat(managesieve): add additional test for usage of spaces as delimiters in SASL PLAIN

chibenwa commented:
Recent changes regarding SASL improvement impact this PR

```
10:42:32,414 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.14.0:compile (default-compile) on project protocols-managesieve: Compilation failure: Compilation failure: 
more_vert
9765	11:42:32 AM	
10:42:32,414 [ERROR] /home/jenkins/workspace/james_ApacheJames_PR-2773/protocols/managesieve/src/main/java/org/apache/james/managesieve/api/Session.java:[27,38] cannot find symbol
more_vert
9766	11:42:32 AM	
10:42:32,414 [ERROR]   symbol:   class OidcSASLConfiguration
more_vert
9767	11:42:32 AM	
10:42:32,414 [ERROR]   location: package org.apache.james.protocols.api
more_vert
9768	11:42:32 AM	
10:42:32,414 [ERROR] /home/jenkins/workspace/james_ApacheJames_PR-2773/protocols/managesieve/src/main/java/org/apache/james/managesieve/api/Session.java:[57,14] cannot find symbol
more_vert
9769	11:42:32 AM	
10:42:32,414 [ERROR]   symbol:   class OidcSASLConfiguration
more_vert
9770	11:42:32 AM	
10:42:32,414 [ERROR]   location: interface org.apache.james.managesieve.api.Session
more_vert
9771	11:42:32 AM	
10:42:32,414 [ERROR] /home/jenkins/workspace/james_ApacheJames_PR-2773/protocols/managesieve/src/main/java/org/apache/james/managesieve/api/Session.java:[59,44] cannot find symbol
more_vert
9772	11:42:32 AM	
10:42:32,414 [ERROR]   symbol:   class OidcSASLConfiguration
more_vert
9773	11:42:32 AM	
10:42:32,414 [ERROR]   location: interface org.apache.james.managesieve.api.Session
more_vert
9774	11:42:32 AM	
10:42:32,414 [ERROR] /home/jenkins/workspace/james_ApacheJames_PR-2773/protocols/managesieve/src/main/java/org/apache/james/managesieve/core/OAUTHAuthenticationProcessor.java:[34,38] cannot find symbol
more_vert
9775	11:42:32 AM	
10:42:32,414 [ERROR]   symbol:   class OidcSASLConfiguration
more_vert
9776	11:42:32 AM	
10:42:32,414 [ERROR]   location: package org.apache.james.protocols.api
more_vert
9777	11:42:32 AM	
10:42:32,414 [ERROR] /home/jenkins/workspace/james_ApacheJames_PR-2773/protocols/managesieve/src/main/java/org/apache/james/managesieve/core/OAUTHAuthenticationProcessor.java:[40,19] cannot find symbol
more_vert
9778	11:42:32 AM	
10:42:32,414 [ERROR]   symbol:   class OidcSASLConfiguration
more_vert
9779	11:42:32 AM	
10:42:32,414 [ERROR]   location: class org.apache.james.managesieve.core.OAUTHAuthenticationProcessor
more_vert
9780	11:42:32 AM	
10:42:32,414 [ERROR] /home/jenkins/workspace/james_ApacheJames_PR-2773/protocols/managesieve/src/main/java/org/apache/james/managesieve/core/OAUTHAuthenticationProcessor.java:[42,41] cannot find symbol
more_vert
9781	11:42:32 AM	
10:42:32,414 [ERROR]   symbol:   class OidcSASLConfiguration
more_vert
9782	11:42:32 AM	
10:42:32,415 [ERROR]   location: class org.apache.james.managesieve.core.OAUTHAuthenticationProcessor
more_vert
9783	11:42:32 AM	
10:42:32,415 [ERROR] /home/jenkins/workspace/james_ApacheJames_PR-2773/protocols/managesieve/src/main/java/org/apache/james/managesieve/util/SettableSession.java:[28,38] cannot find symbol
more_vert
9784	11:42:32 AM	
10:42:32,415 [ERROR]   symbol:   class OidcSASLConfiguration
more_vert
9785	11:42:32 AM	
10:42:32,415 [ERROR]   location: package org.apache.james.protocols.api
more_vert
9786	11:42:32 AM	
10:42:32,415 [ERROR] /home/jenkins/workspace/james_ApacheJames_PR-2773/protocols/managesieve/src/main/java/org/apache/james/managesieve/util/SettableSession.java:[36,22] cannot find symbol
more_vert
9787	11:42:32 AM	
10:42:32,415 [ERROR]   symbol:   class OidcSASLConfiguration
more_vert
9788	11:42:32 AM	
10:42:32,415 [ERROR]   location: class org.apache.james.managesieve.util.SettableSession
more_vert
9789	11:42:32 AM	
10:42:32,415 [ERROR] /home/jenkins/workspace/james_ApacheJames_PR-2773/protocols/managesieve/src/main/java/org/apache/james/managesieve/util/SettableSession.java:[89,21] cannot find symbol
more_vert
9790	11:42:32 AM	
10:42:32,415 [ERROR]   symbol:   class OidcSASLConfiguration
more_vert
9791	11:42:32 AM	
10:42:32,415 [ERROR]   location: class org.apache.james.managesieve.util.SettableSession
more_vert
9792	11:42:32 AM	
10:42:32,415 [ERROR] /home/jenkins/workspace/james_ApacheJames_PR-2773/protocols/managesieve/src/main/java/org/apache/james/managesieve/util/SettableSession.java:[94,51] cannot find symbol
more_vert
9793	11:42:32 AM	
10:42:32,415 [ERROR]   symbol:   class OidcSASLConfiguration
more_vert
9794	11:42:32 AM	
10:42:32,415 [ERROR]   location: class org.apache.james.managesieve.util.SettableSession
more_vert
9795	11:42:32 AM	
10:42:32,415 [ERROR] -> [Help 1]
```

A rebase is needed. If needed I can do it.

felixauringer committed:
refactor(managesieve): rebase on new oidc token validation

felixauringer commented:
I have rebased on the newest commit on main.
If the tests are green, I still have to adapt the documentation (but as far as I can see, this has to be done for everything using the new audience handling).

Before the new `validateToken` method was introduced, I had already written tests in the Managesieve implementation for all different cases regarding introspection, local validation, userinfo. Those tests would be better located in `OidcJwtTokenVerifierTest` now in my opinion. Do you agree? If so, I could move them there (currently there are no tests for the different cases in the config as far as I can tell) and focus more on the Managesieve related functionality in the Managesieve tests.
To keep my old tests working, I also had to adapt your new code a little bit. In `OidcSASLConfiguration`, there were two static variables relying on System Properties. As they are set once in the JVM, I did not find a good way to run tests with different values for them. They are therefore now read when the configuration is parsed. This should not happen often and therefore these additional calls of `System.getProperty` should be fine in my opinion but please correct me if you know better.

chibenwa reviewed (approved):
Sorry for the (very) long awaited review.

Cheers.

Arsnael reviewed (approved):
LGTM. Thanks for this contribution @felixauringer :)

chibenwa commented:
I merged this work as it had been open for a real long time.

Thanks for putting it together @felixauringer 

Other enhancements are likely to be done as follow up PRs if relevant.
